### DISCUSSION TITLE
Applying talib "all_all_ta_features" throws an error in Strategy

### DISCUSSION THREAD

**Discussion by adamjbradley** [kernc/backtesting.py#983] [Q&A] [answered]
Or should I just apply to the df prior to passing it to the Backtest call?
```python


class SprayAndPray(Strategy):
  def init(self):
    def ApplyAllIndicators(data):
      print(data.df)
      data.df = add_all_ta_features(data.df, open="Open", high="High", low="Low", close="Close", volume="Volume", fillna=True)
      return ApplyAllIndicators(data)
  
  self.I(ApplyAllIndicators, self.data)

  def next(self):
  ...

**Comment by kernc** [comment, marked as answer]
> Or should I just apply to the df prior to passing it to the Backtest call?

That would be better. Since it's generated on-the-fly, [`Strategy.data.df`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.data) cannot be assigned to in the way you expect.

**Reply by adamjbradley** [reply]
Thank you!

**Reply by BradKML** [reply]
@adamjbradley as a note from the future, does this include "candle sticks" (medium frequency) or just only momentum (slow trades)?
